californium-core/src/main/java/org/eclipse/californium/core/network/UdpMatcher.java

```diff
@@ -396,11 +398,12 @@ public void completed(final Exchange exchange) {
 
 		@Override
 		public void contextEstablished(final Exchange exchange) {
-
-			if (exchange.getRequest() != null) {
-				observationStore.setContext(exchange.getRequest().getToken(), exchange.getCorrelationContext());
+			Request currentRequest = exchange.getCurrentRequest(); 
+			Request request = exchange.getRequest(); 
+			if (request == currentRequest && request != null && request.isObserve()) {
+				observationStore.setContext(request.getToken(), exchange.getCorrelationContext());
```

Comment:
@sophokles73  read #311, and I'm pretty sure, you can be more precise in what you not understand or where you have a different view.

I created this issue #311 to discuss this topic there even before this PR. This PR was only the part of obvious optimizations. From the available source alone It's hard to understand, what exactly should be achieved with or must be obeyed for the context correlation processing.

To use (null == correlationContext) for further optimizations would only optimize blockwise. I think we need more optimizations here.

So: start reading #311 and discuss there!